src/python/pants/backend/java/package/deploy_jar.py

```diff
@@ -43,15 +50,61 @@ class DeployJarFieldSet(PackageFieldSet):
     dependencies: Dependencies
 
 
+class DeployJarClasspathEntryRequest(ClasspathEntryRequest):
+    field_sets = (DeployJarFieldSet,)
+    # A `deploy_jar` can have a Classpath requested for it, but should not be used as a dependency.
+    root_only = True
+
+
+@rule
+async def deploy_jar_classpath(
+    request: DeployJarClasspathEntryRequest,
+    union_membership: UnionMembership,
+) -> FallibleClasspathEntry:
+    if len(request.component.members) > 1:
+        # If multiple DeployJar targets were coarsened into a single instance, it's because they
+        # formed a cycle among themselves... but at a high level, they shouldn't have dependencies
+        # on one another anyway.
+        raise Exception(
+            "`deploy_jar` targets should not depend on one another:\n"
+            f"{request.component.bullet_list()}"
+        )
+    classpath_entries = FallibleClasspathEntry.if_all_succeeded(
+        await MultiGet(
+            Get(
+                FallibleClasspathEntry,
```

Comment:
Why use `Fallible`?

Reply:
The contract for `ClasspathEntryRequest` is to return `FallibleClasspathEntry`, in order to fail slow and show all compile errors.